File: src/lib/src/main/java/work/lclpnet/ap2/api/base/Participants.java
package work.lclpnet.ap2.api.base;

import net.minecraft.server.level.ServerPlayer;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Stream;

public interface Participants extends Iterable<ServerPlayer> {

    /**
     * @return The currently participating players.
     */
    Set<ServerPlayer> getAsSet();

    void remove(ServerPlayer player);

    boolean isParticipating(UUID uuid);

    @NotNull
    @Override
    default Iterator<ServerPlayer> iterator() {
        return getAsSet().iterator();
    }

    default boolean isParticipating(ServerPlayer player) {
        return isParticipating(player.getUUID());
    }

    default int count() {
        return getAsSet().size();
    }

    default Optional<ServerPlayer> getRandomParticipant(Random random) {
        int count = count();

        if (count <= 0) {
            return Optional.empty();
        }

        return stream().skip(random.nextInt(count)).findFirst();
    }

    default Optional<ServerPlayer> getParticipant(UUID uuid) {
        return stream()
                .filter(player -> player.getUUID().equals(uuid))
                .findAny();
    }

    default Stream<ServerPlayer> stream() {
        return getAsSet().stream();
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/player/Participants.kt
package work.lclpnet.ap2.game.player

import net.minecraft.server.level.ServerPlayer
import work.lclpnet.ap2.game.data.type.PlayerRef
import java.util.*
import java.util.stream.Stream

interface Participants : Iterable<ServerPlayer> {
    /**
     * @return The currently participating players.
     */
    val asSet: Set<ServerPlayer>

    val initialParticipants: Set<PlayerRef>

    fun remove(player: ServerPlayer)

    fun isParticipating(uuid: UUID): Boolean

    override fun iterator(): Iterator<ServerPlayer> = asSet.iterator()

    fun isParticipating(player: ServerPlayer): Boolean =
        isParticipating(player.getUUID())

    fun count(): Int = asSet.size

    fun getRandomParticipant(random: Random): ServerPlayer? {
        val count = count()

        if (count <= 0) {
            return null
        }

        return asSet.randomOrNull()
    }

    fun getParticipant(uuid: UUID): ServerPlayer? =
        firstOrNull { it.uuid == uuid }

    fun stream(): Stream<ServerPlayer> =
        asSet.stream()
}
